cities/queries/causal_insight.py

```diff
@@ -0,0 +1,413 @@
+
+        self.intervened_value_original = revert_standardize_and_scale_scaler(
+            self.intervened_value, self.year, self.intervention_dataset
+        )
+
+        # interventions_this_year = dg.std_wide[self.intervention_dataset][str(year)]
+
+        # nearest_intervention_index = min(
+        #     range(len(interventions_this_year)),
+        #     key=lambda i: abs(interventions_this_year[i] - self.intervened_value),
+        # )
+        # self.intervened_value_original = interventions_this_year_original[
+        #     nearest_intervention_index
+        # ]
+
+        self.intervened_value_percentile = round(
+            (
+                np.mean(
+                    interventions_this_year_original.values
+                    <= self.intervened_value_original
+                )
+                * 100
+            ),
+            1,
+        )
+
+        self.fips_id = (
+            dg.std_wide[self.intervention_dataset]
+            .loc[dg.std_wide[self.intervention_dataset]["GeoFIPS"] == fips]
+            .index[0]
+        )
+
+        self.name = dg.std_wide[self.intervention_dataset]["GeoName"].iloc[self.fips_id]
+
+        # get observed values at the prediction times
+        self.observed_intervention = dg.std_wide[self.intervention_dataset].iloc[
+            self.fips_id
+        ][str(year)]
+
+        self.observed_intervention_original = dg.wide[self.intervention_dataset].iloc[
+            self.fips_id
+        ][str(year)]
+
+        self.observed_outcomes = dg.std_wide[self.outcome_dataset].iloc[self.fips_id][
+            outcome_years[year_id : (year_id + 4)]
+        ]
+        self.intervention_diff = self.intervened_value - self.observed_intervention
+
+        self.intervention_impact = {}
+        self.intervention_impact_mean = []
+        self.intervention_impact_low = []
+        self.intervention_impact_high = []
+        for shift in [1, 2, 3]:
+            self.intervention_impact[shift] = (
+                self.tensed_tau_samples[shift] * self.intervention_diff
+            )
+            self.intervention_impact_mean.append(
+                np.mean(self.intervention_impact[shift])
+            )
+            self.intervention_impact_low.append(
+                np.percentile(self.intervention_impact[shift], 2.5)
+            )
+            self.intervention_impact_high.append(
+                np.percentile(self.intervention_impact[shift], 97.5)
+            )
+
+        predicted_mean = [self.observed_outcomes.iloc[0]] + (
+            self.intervention_impact_mean + self.observed_outcomes.iloc[1:]
+        ).tolist()
+        predicted_low = [self.observed_outcomes.iloc[0]] + (
+            self.intervention_impact_low + self.observed_outcomes.iloc[1:]
+        ).tolist()
+        predicted_high = [self.observed_outcomes.iloc[0]] + (
+            self.intervention_impact_high + self.observed_outcomes.iloc[1:]
+        ).tolist()
+
+        self.predictions = pd.DataFrame(
```

Comment:
Went through the algebra just to confirm my intuition that this is indeed just the counterfactual — checks out to me:

obs = bias + f(X, Y) + tau * T + eps
eps = obs - (bias + f(X, Y) + tau * T)
int = bias + f(X, Y) + tau * T' + eps

int = bias + f(X, Y) + tau * T’ + obs - (bias + f(X, Y) + tau * T)
= tau * T’ + obs - tau * T
= tau * ( T’ - T) + obs

Reply:
A good sanity check here might be do run the full interventional predictive (not counterfactual) and make sure that the counterfactual computed here sits within the interventional predictive (where noise isn't assumed to be shared between the two regimes).

Reply:
Thanks for the sanity check here! As for running full predictive, the posterior is so wide that it's not hard to sit within. :)

Reply:
Ah I see. I guess that's true statistically, but it would still catch serious bugs (though I think we'd have seen their effect elsewhere). Anyway — could still be useful down the road.